### ISSUE TITLE
Implicit dependencies based on user mount options are incorrectly configured for mount units activated by external execution of mount command.

### ISSUE THREAD

**Issue by d-hatayama** [systemd/systemd#23552]
**systemd version the issue has been seen with**

```
# rpm -q systemd
systemd-250.6-1.fc36.x86_64
# /usr/lib/systemd/systemd --version
systemd 250 (v250.6-1.fc36)
+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
```

**Used distribution**

Fedora Linux 36

**Linux kernel version used** (`uname -a`)

```
# uname -a
Linux localhost.localdomain 5.17.11-300.fc36.x86_64 #1 SMP PREEMPT Wed May 25 15:04:05 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

Other relevant packages:

```
# rpm -q util-linux libmount
util-linux-2.38-0.2.fc36.x86_64
libmount-2.38-0.2.fc36.x86_64
```

**CPU architecture issue was seen on**

```
# arch
x86_64
```

**Expected behaviour you didn't see**

Implicit dependencies based on user mount options are correctly configured for mount units activated by external execution of mount command.

**Unexpected behaviour you saw**

Implicit dependencies based on user mount options are incorrectly configured for mount units activated by external execution of mount command.

For example, `Before=local-fs.target` is unexpectedly configured although `_netdev` mount option is configured.

**Steps to reproduce the problem**

Let's see, due to this issue, that `opt.mount` gets `Before=local-fs.target` although `_netdev` mount option is configured.

Prepare an file system you can make use of for this step in advance. Here `/dev/vdb1` is used.

Use the following gdb python script `repro.py`:

```python:repro.py
import gdb
import time

gdb.execute('file /usr/bin/mount')
# 82 sys_rename
gdb.execute('catch syscall 82')
gdb.execute('run -o _netdev /dev/vdb1 /opt')
time.sleep(10)
# Two continue commands because break occurs not only at the entry of rename() but also at the return.
gdb.execute('continue')
gdb.execute('continue')
gdb.execute('quit')
```

Run it:

```
# gdb --batch -x ./repro.py
```

Then, `opt.mount` has `Before=local-fs.target` although `_netdev` mount option is configured.

```
# systemctl show -p Before,After opt.mount
Before=umount.target remote-fs.target local-fs.target                                                 
After=blockdev@dev-vdb1.target local-fs-pre.target network.target network-online.target dev-vdb1.device remote-fs-pre.target systemd-journald.socket
```

This also results in ordering cycle at system shutdown in combination with `After=network.target` configured via `_netdev` mount option like:

```
May 29 20:14:09 localhost.localdomain systemd[1]: basic.target: Found ordering cycle on NetworkManager.service/stop
May 29 20:14:09 localhost.localdomain systemd[1]: basic.target: Found dependency on network.target/stop
May 29 20:14:09 localhost.localdomain systemd[1]: basic.target: Found dependency on opt.mount/stop
May 29 20:14:09 localhost.localdomain systemd[1]: basic.target: Found dependency on local-fs.target/stop
May 29 20:14:09 localhost.localdomain systemd[1]: basic.target: Found dependency on auditd.service/stop
May 29 20:14:09 localhost.localdomain systemd[1]: basic.target: Found dependency on sysinit.target/stop
May 29 20:14:09 localhost.localdomain systemd[1]: basic.target: Found dependency on basic.target/stop
May 29 20:14:09 localhost.localdomain systemd[1]: basic.target: Job NetworkManager.service/stop deleted to break ordering cycle starting with basic.target/stop
```

**Additional program output to the terminal or log subsystem illustrating the issue**

`mount` command passes user mount options via `/run/mount/utab` file. `mount()` system call is first executed and then `/run/mount/utab` file is updated.

```
# LANG=C strace -s 1024 -e mount,openat,write,rename mount -o _netdev /dev/vdb1 /opt
...snip...
mount("/dev/vdb1", "/opt", "xfs", 0, NULL) = 0
openat(AT_FDCWD, "/run/mount/utab.lock", O_RDONLY|O_CREAT|O_CLOEXEC, 0644) = 3
openat(AT_FDCWD, "/run/mount/utab", O_RDONLY|O_CLOEXEC) = 4
openat(AT_FDCWD, "/run/mount/utab.krbhsX", O_RDWR|O_CREAT|O_EXCL|O_CLOEXEC, 0600) = 4
write(4, "SRC=/dev/vdb1 TARGET=/opt ROOT=/ OPTS=_netdev\n", 46) = 46
rename("/run/mount/utab.krbhsX", "/run/mount/utab") = 0
+++ exited with 0 +++
```

Thus, systemd cannot handle user mount options correctly if systemd opens `/run/mount/utab` file before `mount` command completes renaming `/run/mount/utab` file.

**Comment by yuwata** [comment]
I guess this is a duplicate of #19983.

**Comment by yuwata** [comment]
If possible, could you try #23367?

**Comment by d-hatayama** [comment]
Thanks. I tried #23367 and opt.mount never gets Before=local-fs.target in the above reproduction steps.

On the other hand, if I have good understanding, even with the fix #23367, Before=local-fs.target is still configured temporarily until it is removed at the update of /run/mount/utab. If so, it means that wrong configuration is configured temporarily, which is still problematic behavior. Is it possible to move configuration of dependencies to after the update of /run/mount/utab?

**Comment by d-hatayama** [comment]
@yuwata Could you review PR #23929? This patch series aims to fix this issue, addressing the concern I mentioned at https://github.com/systemd/systemd/issues/23552#issuecomment-1143410198. Is it possible to handle the issue here separately to #19983?

**Comment by yuwata** [comment]
@d-hatayama Note, thanks to your PR, now I understand this issue seems different from #19983. I will try to address this after #23929 or other PRs about .mount or .device unit handling are merged. Of course, anytime, another PR (or any suggestions) for this issue is welcome.
